In 2.5.0 this worked like a charm, now I have some troubles, I think they come from 982fc526c04e615e3d83a27bddc3629e36efb978.

The cruxial point is, how can I use `force_in` for integer degree celsius to degree fahrenheit? @mpusz if you can give me a hint, I can continue on the PR.

<details>
<summary>Lengthy compiler errors</summary>
```
In file included from ../../src/core/include/mp-units/math.h:29,
                 from ../../test/runtime/math_test.cpp:34:
../../src/core/include/mp-units/framework/quantity_point.h: In instantiation of ‘constexpr mp_units::quantity_point<R, PO, Rep>::quantity_point(const QP&) [with QP = mp_units::quantity_point<mp_units::usc::degree_Fahrenheit(), mp_units::usc::fahrenheit_zero(), int>; auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); auto [requires mp_units::PointOriginFor<<placeholder>, get_quantity_spec(R)>] PO = mp_units::si::ice_point(); Rep = int]’:
../../src/core/include/mp-units/math.h:371:12:   required from here
  371 |     return res;
      |            ^~~
../../test/runtime/math_test.cpp:406:7:   in ‘constexpr’ expansion of ‘mp_units::floor<usc::degree_Fahrenheit(), si::deci_<si::degree_Celsius>(), si::ice_point(), int>(mp_units::point<mp_units::si::deci_<mp_units::si::degree_Celsius>()>.mp_units::point_<mp_units::si::deci_<mp_units::si::degree_Celsius> >::operator()<int>(-125))’
../../src/core/include/mp-units/framework/quantity_point.h:587:49: error: no matching function for call to ‘mp_units::quantity<mp_units::usc::degree_Fahrenheit(), int>::quantity(<brace-enclosed initializer list>)’
  587 |           return quantity_type{qp - point_origin};
      |                                                 ^
../../src/core/include/mp-units/framework/quantity_point.h:587:49: note: there are 12 candidates
In file included from ../../test/runtime/almost_equals.h:30,
                 from ../../test/runtime/math_test.cpp:23:
../../src/core/include/mp-units/framework/quantity.h:553:85: note: candidate 1: ‘template<class Q>  requires (QuantityLike<Q>) && (QuantityConstructibleFrom<mp_units::quantity<R, Rep>, mp_units::quantity<mp_units::quantity_like_traits<Q>::reference, typename mp_units::quantity_like_traits<Q>::rep> >) constexpr mp_units::quantity<R, Rep>::quantity(const Q&) [with auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); Rep = int]’
  553 |                      !std::convertible_to<detail::quantity_like_type<Q>, quantity>) quantity(const Q& q) :
      |                                                                                     ^~~~~~~~
../../src/core/include/mp-units/framework/quantity.h:553:85: note: template argument deduction/substitution failed:
../../src/core/include/mp-units/framework/quantity.h:553:85: note: constraints not satisfied
In file included from ../../src/core/include/mp-units/bits/sudo_cast.h:26,
                 from ../../src/core/include/mp-units/framework/quantity.h:28:
../../src/core/include/mp-units/framework/quantity_concepts.h: In substitution of ‘template<class Q>  requires (QuantityLike<Q>) && (QuantityConstructibleFrom<mp_units::quantity<R, Rep>, mp_units::quantity<mp_units::quantity_like_traits<Q>::reference, typename mp_units::quantity_like_traits<Q>::rep> >) constexpr mp_units::quantity<mp_units::usc::degree_Fahrenheit(), int>::quantity(const Q&) [with Q = mp_units::quantity<mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >(), int>]’:
../../src/core/include/mp-units/framework/quantity_point.h:587:49:   required from ‘constexpr mp_units::quantity_point<R, PO, Rep>::quantity_point(const QP&) [with QP = mp_units::quantity_point<mp_units::usc::degree_Fahrenheit(), mp_units::usc::fahrenheit_zero(), int>; auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); auto [requires mp_units::PointOriginFor<<placeholder>, get_quantity_spec(R)>] PO = mp_units::si::ice_point(); Rep = int]’
  587 |           return quantity_type{qp - point_origin};
      |                                                 ^
../../src/core/include/mp-units/math.h:371:12:   required from here
  371 |     return res;
      |            ^~~
../../src/core/include/mp-units/framework/quantity_concepts.h:86:9:   required for the satisfaction of ‘QuantityLike<Q>’ [with Q = mp_units::quantity<mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}, int>]
../../src/core/include/mp-units/framework/quantity_concepts.h:86:24: note: the expression ‘!(Quantity<T>) [with T = mp_units::quantity<mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}, int>]’ evaluated to ‘false’
   86 | concept QuantityLike = !Quantity<T> && detail::QuantityLikeImpl<T, quantity_like_traits> && requires {
      |                        ^~~~~~~~~~~~
../../src/core/include/mp-units/framework/quantity_point.h: In instantiation of ‘constexpr mp_units::quantity_point<R, PO, Rep>::quantity_point(const QP&) [with QP = mp_units::quantity_point<mp_units::usc::degree_Fahrenheit(), mp_units::usc::fahrenheit_zero(), int>; auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); auto [requires mp_units::PointOriginFor<<placeholder>, get_quantity_spec(R)>] PO = mp_units::si::ice_point(); Rep = int]’:
../../src/core/include/mp-units/math.h:371:12:   required from here
  371 |     return res;
      |            ^~~
../../src/core/include/mp-units/framework/quantity.h:544:5: note: candidate 2: ‘template<auto R2, class Rep2>  requires (QuantityConstructibleFrom<mp_units::quantity<R, Rep>, mp_units::quantity<R2, Rep2> >) && !equivalent(mp_units::quantity<R, Rep>::unit, get_unit(R2)) constexpr mp_units::quantity<R, Rep>::quantity(const mp_units::quantity<R2, Rep2>&) [with auto R2 = R2; auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); Rep = int]’
  544 |     quantity(const quantity<R2, Rep2>& q) :
      |     ^~~~~~~~
../../src/core/include/mp-units/framework/quantity.h:544:5: note: template argument deduction/substitution failed:
../../src/core/include/mp-units/framework/quantity.h:544:5: note: constraints not satisfied
../../src/core/include/mp-units/framework/quantity_point.h: In substitution of ‘template<auto R2, class Rep2>  requires (QuantityConstructibleFrom<mp_units::quantity<R, Rep>, mp_units::quantity<R2, Rep2> >) && !equivalent(mp_units::quantity<R, Rep>::unit, get_unit(R2)) constexpr mp_units::quantity<mp_units::usc::degree_Fahrenheit(), int>::quantity(const mp_units::quantity<R, Rep>&) [with auto R2 = mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >(); Rep2 = int]’:
../../src/core/include/mp-units/framework/quantity_point.h:587:49:   required from ‘constexpr mp_units::quantity_point<R, PO, Rep>::quantity_point(const QP&) [with QP = mp_units::quantity_point<mp_units::usc::degree_Fahrenheit(), mp_units::usc::fahrenheit_zero(), int>; auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); auto [requires mp_units::PointOriginFor<<placeholder>, get_quantity_spec(R)>] PO = mp_units::si::ice_point(); Rep = int]’
  587 |           return quantity_type{qp - point_origin};
      |                                                 ^
../../src/core/include/mp-units/math.h:371:12:   required from here
  371 |     return res;
      |            ^~~
../../test/runtime/math_test.cpp:406:7:   in ‘constexpr’ expansion of ‘mp_units::floor<usc::degree_Fahrenheit(), si::deci_<si::degree_Celsius>(), si::ice_point(), int>(mp_units::point<mp_units::si::deci_<mp_units::si::degree_Celsius>()>.mp_units::point_<mp_units::si::deci_<mp_units::si::degree_Celsius> >::operator()<int>(-125))’
../../src/core/include/mp-units/framework/quantity.h:99:9:   required for the satisfaction of ‘ImplicitConversion<QFrom::unit, typename QFrom::rep, QTo::unit, typename QTo::rep>’ [with QFrom = mp_units::quantity<mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}, int>; QTo = mp_units::quantity<mp_units::usc::degree_Fahrenheit{}, int>]
../../src/core/include/mp-units/framework/quantity.h:105:9:   required for the satisfaction of ‘QuantityConstructibleFrom<mp_units::quantity<R, Rep>, mp_units::quantity<R2, Rep2> >’ [with R = mp_units::usc::degree_Fahrenheit{}; Rep = int; R2 = mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}; Rep2 = int]
../../src/core/include/mp-units/framework/quantity.h:100:95: note: no operand of the disjunction is satisfied
  100 |                              (mp_units::implicitly_scalable<FromUnit, FromRep, ToUnit, ToRep> ||
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
  101 |                               unsatisfied<"Conversion from '{}' as '{}' to '{}' as '{}' is truncating">(
      |                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  102 |                                 unit_symbol(FromUnit), type_name<FromRep>(), unit_symbol(ToUnit), type_name<ToRep>()));
      |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/core/include/mp-units/framework/quantity.h:100:41: note: the operand ‘implicitly_scalable<FromUnit, FromRep, ToUnit, ToRep>’ is unsatisfied because
  100 |                              (mp_units::implicitly_scalable<FromUnit, FromRep, ToUnit, ToRep> ||
      |                              ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  101 |                               unsatisfied<"Conversion from '{}' as '{}' to '{}' as '{}' is truncating">(
      |                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  102 |                                 unit_symbol(FromUnit), type_name<FromRep>(), unit_symbol(ToUnit), type_name<ToRep>()));
      |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/core/include/mp-units/framework/quantity.h:99:9:   required for the satisfaction of ‘ImplicitConversion<QFrom::unit, typename QFrom::rep, QTo::unit, typename QTo::rep>’ [with QFrom = mp_units::quantity<mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}, int>; QTo = mp_units::quantity<mp_units::usc::degree_Fahrenheit{}, int>]
../../src/core/include/mp-units/framework/quantity.h:105:9:   required for the satisfaction of ‘QuantityConstructibleFrom<mp_units::quantity<R, Rep>, mp_units::quantity<R2, Rep2> >’ [with R = mp_units::usc::degree_Fahrenheit{}; Rep = int; R2 = mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}; Rep2 = int]
../../src/core/include/mp-units/framework/quantity.h:100:41: note: the expression ‘implicitly_scalable<FromUnit, FromRep, ToUnit, ToRep> [with FromUnit = mp_units::quantity<mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}, int>::unit; FromRep = int; ToUnit = mp_units::quantity<mp_units::usc::degree_Fahrenheit{}, int>::unit; ToRep = int]’ evaluated to ‘false’
  100 |                              (mp_units::implicitly_scalable<FromUnit, FromRep, ToUnit, ToRep> ||
      |                               ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/core/include/mp-units/framework/quantity.h:101:104: note: the operand ‘unsatisfied<"Conversion from \'{}\' as \'{}\' to \'{}\' as \'{}\' is truncating">(mp_units::unit_symbol(FromUnit), type_name<FromRep>(), mp_units::unit_symbol(ToUnit), type_name<ToRep>())’ is unsatisfied because
  100 |                              (mp_units::implicitly_scalable<FromUnit, FromRep, ToUnit, ToRep> ||
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~        
  101 |                               unsatisfied<"Conversion from '{}' as '{}' to '{}' as '{}' is truncating">(
      |                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
  102 |                                 unit_symbol(FromUnit), type_name<FromRep>(), unit_symbol(ToUnit), type_name<ToRep>()));
      |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../test/runtime/math_test.cpp:406:7:   in ‘constexpr’ expansion of ‘mp_units::floor<usc::degree_Fahrenheit(), si::deci_<si::degree_Celsius>(), si::ice_point(), int>(mp_units::point<mp_units::si::deci_<mp_units::si::degree_Celsius>()>.mp_units::point_<mp_units::si::deci_<mp_units::si::degree_Celsius> >::operator()<int>(-125))’
../../src/core/include/mp-units/framework/quantity.h:99:9:   required for the satisfaction of ‘ImplicitConversion<QFrom::unit, typename QFrom::rep, QTo::unit, typename QTo::rep>’ [with QFrom = mp_units::quantity<mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}, int>; QTo = mp_units::quantity<mp_units::usc::degree_Fahrenheit{}, int>]
../../src/core/include/mp-units/framework/quantity.h:105:9:   required for the satisfaction of ‘QuantityConstructibleFrom<mp_units::quantity<R, Rep>, mp_units::quantity<R2, Rep2> >’ [with R = mp_units::usc::degree_Fahrenheit{}; Rep = int; R2 = mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}; Rep2 = int]
../../src/core/include/mp-units/framework/quantity.h:101:104: note: the expression ‘(unsatisfied<"Conversion from \'{}\' as \'{}\' to \'{}\' as \'{}\' is truncating">)(mp_units::unit_symbol(FromUnit), (type_name<FromRep>)(), mp_units::unit_symbol(ToUnit), (type_name<ToRep>)()) [with FromUnit = mp_units::quantity<mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}, int>::unit; FromRep = int; ToUnit = mp_units::quantity<mp_units::usc::degree_Fahrenheit{}, int>::unit; ToRep = int]’ evaluated to ‘false’
  101 |                               unsatisfied<"Conversion from '{}' as '{}' to '{}' as '{}' is truncating">(
      |                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
  102 |                                 unit_symbol(FromUnit), type_name<FromRep>(), unit_symbol(ToUnit), type_name<ToRep>()));
      |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/core/include/mp-units/framework/quantity_point.h: In instantiation of ‘constexpr mp_units::quantity_point<R, PO, Rep>::quantity_point(const QP&) [with QP = mp_units::quantity_point<mp_units::usc::degree_Fahrenheit(), mp_units::usc::fahrenheit_zero(), int>; auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); auto [requires mp_units::PointOriginFor<<placeholder>, get_quantity_spec(R)>] PO = mp_units::si::ice_point(); Rep = int]’:
../../src/core/include/mp-units/math.h:371:12:   required from here
  371 |     return res;
      |            ^~~
../../src/core/include/mp-units/framework/quantity.h:533:5: note: candidate 3: ‘template<auto R2, class Rep2>  requires (QuantityConstructibleFrom<mp_units::quantity<R, Rep>, mp_units::quantity<R2, Rep2> >) && equivalent(mp_units::quantity<R, Rep>::unit, get_unit(R2)) constexpr mp_units::quantity<R, Rep>::quantity(const mp_units::quantity<R2, Rep2>&) [with auto R2 = R2; auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); Rep = int]’
  533 |     quantity(const quantity<R2, Rep2>& q) :
      |     ^~~~~~~~
../../src/core/include/mp-units/framework/quantity.h:533:5: note: template argument deduction/substitution failed:
../../src/core/include/mp-units/framework/quantity.h:533:5: note: constraints not satisfied
../../src/core/include/mp-units/framework/quantity_point.h: In substitution of ‘template<auto R2, class Rep2>  requires (QuantityConstructibleFrom<mp_units::quantity<R, Rep>, mp_units::quantity<R2, Rep2> >) && equivalent(mp_units::quantity<R, Rep>::unit, get_unit(R2)) constexpr mp_units::quantity<mp_units::usc::degree_Fahrenheit(), int>::quantity(const mp_units::quantity<R, Rep>&) [with auto R2 = mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >(); Rep2 = int]’:
../../src/core/include/mp-units/framework/quantity_point.h:587:49:   required from ‘constexpr mp_units::quantity_point<R, PO, Rep>::quantity_point(const QP&) [with QP = mp_units::quantity_point<mp_units::usc::degree_Fahrenheit(), mp_units::usc::fahrenheit_zero(), int>; auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); auto [requires mp_units::PointOriginFor<<placeholder>, get_quantity_spec(R)>] PO = mp_units::si::ice_point(); Rep = int]’
  587 |           return quantity_type{qp - point_origin};
      |                                                 ^
../../src/core/include/mp-units/math.h:371:12:   required from here
  371 |     return res;
      |            ^~~
../../test/runtime/math_test.cpp:406:7:   in ‘constexpr’ expansion of ‘mp_units::floor<usc::degree_Fahrenheit(), si::deci_<si::degree_Celsius>(), si::ice_point(), int>(mp_units::point<mp_units::si::deci_<mp_units::si::degree_Celsius>()>.mp_units::point_<mp_units::si::deci_<mp_units::si::degree_Celsius> >::operator()<int>(-125))’
../../src/core/include/mp-units/framework/quantity.h:99:9:   required for the satisfaction of ‘ImplicitConversion<QFrom::unit, typename QFrom::rep, QTo::unit, typename QTo::rep>’ [with QFrom = mp_units::quantity<mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}, int>; QTo = mp_units::quantity<mp_units::usc::degree_Fahrenheit{}, int>]
../../src/core/include/mp-units/framework/quantity.h:105:9:   required for the satisfaction of ‘QuantityConstructibleFrom<mp_units::quantity<R, Rep>, mp_units::quantity<R2, Rep2> >’ [with R = mp_units::usc::degree_Fahrenheit{}; Rep = int; R2 = mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}; Rep2 = int]
../../src/core/include/mp-units/framework/quantity.h:100:95: note: no operand of the disjunction is satisfied
  100 |                              (mp_units::implicitly_scalable<FromUnit, FromRep, ToUnit, ToRep> ||
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
  101 |                               unsatisfied<"Conversion from '{}' as '{}' to '{}' as '{}' is truncating">(
      |                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  102 |                                 unit_symbol(FromUnit), type_name<FromRep>(), unit_symbol(ToUnit), type_name<ToRep>()));
      |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/core/include/mp-units/framework/quantity.h:100:41: note: the operand ‘implicitly_scalable<FromUnit, FromRep, ToUnit, ToRep>’ is unsatisfied because
  100 |                              (mp_units::implicitly_scalable<FromUnit, FromRep, ToUnit, ToRep> ||
      |                              ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  101 |                               unsatisfied<"Conversion from '{}' as '{}' to '{}' as '{}' is truncating">(
      |                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  102 |                                 unit_symbol(FromUnit), type_name<FromRep>(), unit_symbol(ToUnit), type_name<ToRep>()));
      |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/core/include/mp-units/framework/quantity.h:99:9:   required for the satisfaction of ‘ImplicitConversion<QFrom::unit, typename QFrom::rep, QTo::unit, typename QTo::rep>’ [with QFrom = mp_units::quantity<mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}, int>; QTo = mp_units::quantity<mp_units::usc::degree_Fahrenheit{}, int>]
../../src/core/include/mp-units/framework/quantity.h:105:9:   required for the satisfaction of ‘QuantityConstructibleFrom<mp_units::quantity<R, Rep>, mp_units::quantity<R2, Rep2> >’ [with R = mp_units::usc::degree_Fahrenheit{}; Rep = int; R2 = mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}; Rep2 = int]
../../src/core/include/mp-units/framework/quantity.h:100:41: note: the expression ‘implicitly_scalable<FromUnit, FromRep, ToUnit, ToRep> [with FromUnit = mp_units::quantity<mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}, int>::unit; FromRep = int; ToUnit = mp_units::quantity<mp_units::usc::degree_Fahrenheit{}, int>::unit; ToRep = int]’ evaluated to ‘false’
  100 |                              (mp_units::implicitly_scalable<FromUnit, FromRep, ToUnit, ToRep> ||
      |                               ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/core/include/mp-units/framework/quantity.h:101:104: note: the operand ‘unsatisfied<"Conversion from \'{}\' as \'{}\' to \'{}\' as \'{}\' is truncating">(mp_units::unit_symbol(FromUnit), type_name<FromRep>(), mp_units::unit_symbol(ToUnit), type_name<ToRep>())’ is unsatisfied because
  100 |                              (mp_units::implicitly_scalable<FromUnit, FromRep, ToUnit, ToRep> ||
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~        
  101 |                               unsatisfied<"Conversion from '{}' as '{}' to '{}' as '{}' is truncating">(
      |                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
  102 |                                 unit_symbol(FromUnit), type_name<FromRep>(), unit_symbol(ToUnit), type_name<ToRep>()));
      |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../test/runtime/math_test.cpp:406:7:   in ‘constexpr’ expansion of ‘mp_units::floor<usc::degree_Fahrenheit(), si::deci_<si::degree_Celsius>(), si::ice_point(), int>(mp_units::point<mp_units::si::deci_<mp_units::si::degree_Celsius>()>.mp_units::point_<mp_units::si::deci_<mp_units::si::degree_Celsius> >::operator()<int>(-125))’
../../src/core/include/mp-units/framework/quantity.h:99:9:   required for the satisfaction of ‘ImplicitConversion<QFrom::unit, typename QFrom::rep, QTo::unit, typename QTo::rep>’ [with QFrom = mp_units::quantity<mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}, int>; QTo = mp_units::quantity<mp_units::usc::degree_Fahrenheit{}, int>]
../../src/core/include/mp-units/framework/quantity.h:105:9:   required for the satisfaction of ‘QuantityConstructibleFrom<mp_units::quantity<R, Rep>, mp_units::quantity<R2, Rep2> >’ [with R = mp_units::usc::degree_Fahrenheit{}; Rep = int; R2 = mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}; Rep2 = int]
../../src/core/include/mp-units/framework/quantity.h:101:104: note: the expression ‘(unsatisfied<"Conversion from \'{}\' as \'{}\' to \'{}\' as \'{}\' is truncating">)(mp_units::unit_symbol(FromUnit), (type_name<FromRep>)(), mp_units::unit_symbol(ToUnit), (type_name<ToRep>)()) [with FromUnit = mp_units::quantity<mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >{}, int>::unit; FromRep = int; ToUnit = mp_units::quantity<mp_units::usc::degree_Fahrenheit{}, int>::unit; ToRep = int]’ evaluated to ‘false’
  101 |                               unsatisfied<"Conversion from '{}' as '{}' to '{}' as '{}' is truncating">(
      |                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
  102 |                                 unit_symbol(FromUnit), type_name<FromRep>(), unit_symbol(ToUnit), type_name<ToRep>()));
      |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../src/core/include/mp-units/framework/quantity_point.h: In instantiation of ‘constexpr mp_units::quantity_point<R, PO, Rep>::quantity_point(const QP&) [with QP = mp_units::quantity_point<mp_units::usc::degree_Fahrenheit(), mp_units::usc::fahrenheit_zero(), int>; auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); auto [requires mp_units::PointOriginFor<<placeholder>, get_quantity_spec(R)>] PO = mp_units::si::ice_point(); Rep = int]’:
../../src/core/include/mp-units/math.h:371:12:   required from here
  371 |     return res;
      |            ^~~
../../src/core/include/mp-units/framework/quantity.h:520:87: note: candidate 4: ‘template<class Value>  requires (ExplicitFromNumber<mp_units::quantity<R, Rep>::reference>) && !(RepConstructibleFrom<Rep, Value>) constexpr mp_units::quantity<R, Rep>::quantity(Value) [with auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); Rep = int]’ (deleted)
  520 |                      !mp_units::implicitly_convertible(quantity_spec, dimensionless)) quantity(Value val)
      |                                                                                       ^~~~~~~~
../../src/core/include/mp-units/framework/quantity.h:520:87: note: template argument deduction/substitution failed:
../../src/core/include/mp-units/framework/quantity.h:520:87: note: constraints not satisfied
../../src/core/include/mp-units/framework/quantity.h: In substitution of ‘template<class Value>  requires (ExplicitFromNumber<mp_units::quantity<R, Rep>::reference>) && !(RepConstructibleFrom<Rep, Value>) constexpr mp_units::quantity<mp_units::usc::degree_Fahrenheit(), int>::quantity(Value) [with Value = mp_units::quantity<mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >(), int>]’:
../../src/core/include/mp-units/framework/quantity_point.h:587:49:   required from ‘constexpr mp_units::quantity_point<R, PO, Rep>::quantity_point(const QP&) [with QP = mp_units::quantity_point<mp_units::usc::degree_Fahrenheit(), mp_units::usc::fahrenheit_zero(), int>; auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); auto [requires mp_units::PointOriginFor<<placeholder>, get_quantity_spec(R)>] PO = mp_units::si::ice_point(); Rep = int]’
  587 |           return quantity_type{qp - point_origin};
      |                                                 ^
../../src/core/include/mp-units/math.h:371:12:   required from here
  371 |     return res;
      |            ^~~
../../test/runtime/math_test.cpp:406:7:   in ‘constexpr’ expansion of ‘mp_units::floor<usc::degree_Fahrenheit(), si::deci_<si::degree_Celsius>(), si::ice_point(), int>(mp_units::point<mp_units::si::deci_<mp_units::si::degree_Celsius>()>.mp_units::point_<mp_units::si::deci_<mp_units::si::degree_Celsius> >::operator()<int>(-125))’
../../src/core/include/mp-units/framework/quantity.h:118:9:   required for the satisfaction of ‘UnitOne<R>’ [with R = mp_units::quantity<mp_units::usc::degree_Fahrenheit{}, int>::reference]
../../src/core/include/mp-units/framework/quantity.h:122:9:   required for the satisfaction of ‘ExplicitFromNumber<reference>’ [with R = mp_units::usc::degree_Fahrenheit{}; Rep = int]
../../src/core/include/mp-units/framework/quantity.h:119:30: note: the expression ‘equivalent(get_unit(R), mp_units::one) [with R = mp_units::quantity<mp_units::usc::degree_Fahrenheit{}, int>::reference]’ evaluated to ‘false’
  119 |                   (equivalent(get_unit(R), one) && detail::get_associated_quantity(get_unit(R)) == dimensionless);
      |                    ~~~~~~~~~~^~~~~~~~~~~~~~~~~~
../../src/core/include/mp-units/framework/quantity_point.h: In instantiation of ‘constexpr mp_units::quantity_point<R, PO, Rep>::quantity_point(const QP&) [with QP = mp_units::quantity_point<mp_units::usc::degree_Fahrenheit(), mp_units::usc::fahrenheit_zero(), int>; auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); auto [requires mp_units::PointOriginFor<<placeholder>, get_quantity_spec(R)>] PO = mp_units::si::ice_point(); Rep = int]’:
../../src/core/include/mp-units/math.h:371:12:   required from here
  371 |     return res;
      |            ^~~
../../src/core/include/mp-units/framework/quantity.h:513:22: note: candidate 5: ‘template<class Value>  requires (ExplicitFromNumber<mp_units::quantity<R, Rep>::reference>) && (RepConstructibleFrom<Rep, Value>) && !(convertible_to<Value, Rep>) constexpr mp_units::quantity<R, Rep>::quantity(Value) [with auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); Rep = int]’
  513 |   constexpr explicit quantity(Value val) : numerical_value_is_an_implementation_detail_(std::move(val))
      |                      ^~~~~~~~
../../src/core/include/mp-units/framework/quantity.h:513:22: note: template argument deduction/substitution failed:
../../src/core/include/mp-units/framework/quantity.h:513:22: note: constraints not satisfied
../../src/core/include/mp-units/framework/quantity.h: In substitution of ‘template<class Value>  requires (ExplicitFromNumber<mp_units::quantity<R, Rep>::reference>) && (RepConstructibleFrom<Rep, Value>) && !(convertible_to<Value, Rep>) constexpr mp_units::quantity<mp_units::usc::degree_Fahrenheit(), int>::quantity(Value) [with Value = mp_units::quantity<mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >(), int>]’:
../../src/core/include/mp-units/framework/quantity_point.h:587:49:   required from ‘constexpr mp_units::quantity_point<R, PO, Rep>::quantity_point(const QP&) [with QP = mp_units::quantity_point<mp_units::usc::degree_Fahrenheit(), mp_units::usc::fahrenheit_zero(), int>; auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); auto [requires mp_units::PointOriginFor<<placeholder>, get_quantity_spec(R)>] PO = mp_units::si::ice_point(); Rep = int]’
  587 |           return quantity_type{qp - point_origin};
      |                                                 ^
../../src/core/include/mp-units/math.h:371:12:   required from here
  371 |     return res;
      |            ^~~
../../test/runtime/math_test.cpp:406:7:   in ‘constexpr’ expansion of ‘mp_units::floor<usc::degree_Fahrenheit(), si::deci_<si::degree_Celsius>(), si::ice_point(), int>(mp_units::point<mp_units::si::deci_<mp_units::si::degree_Celsius>()>.mp_units::point_<mp_units::si::deci_<mp_units::si::degree_Celsius> >::operator()<int>(-125))’
../../src/core/include/mp-units/framework/quantity.h:118:9:   required for the satisfaction of ‘UnitOne<R>’ [with R = mp_units::quantity<mp_units::usc::degree_Fahrenheit{}, int>::reference]
../../src/core/include/mp-units/framework/quantity.h:122:9:   required for the satisfaction of ‘ExplicitFromNumber<reference>’ [with R = mp_units::usc::degree_Fahrenheit{}; Rep = int]
../../src/core/include/mp-units/framework/quantity.h:119:30: note: the expression ‘equivalent(get_unit(R), mp_units::one) [with R = mp_units::quantity<mp_units::usc::degree_Fahrenheit{}, int>::reference]’ evaluated to ‘false’
  119 |                   (equivalent(get_unit(R), one) && detail::get_associated_quantity(get_unit(R)) == dimensionless);
      |                    ~~~~~~~~~~^~~~~~~~~~~~~~~~~~
../../src/core/include/mp-units/framework/quantity_point.h: In instantiation of ‘constexpr mp_units::quantity_point<R, PO, Rep>::quantity_point(const QP&) [with QP = mp_units::quantity_point<mp_units::usc::degree_Fahrenheit(), mp_units::usc::fahrenheit_zero(), int>; auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); auto [requires mp_units::PointOriginFor<<placeholder>, get_quantity_spec(R)>] PO = mp_units::si::ice_point(); Rep = int]’:
../../src/core/include/mp-units/math.h:371:12:   required from here
  371 |     return res;
      |            ^~~
../../src/core/include/mp-units/framework/quantity.h:500:13: note: candidate 6: ‘template<class FwdValue, class R2>  requires (Reference<R2>) && (!equivalent(mp_units::quantity<R, Rep>::unit, get_unit({})) && (QuantityConstructibleFrom<mp_units::quantity<R, Rep>, mp_units::quantity<{}, typename std::remove_cvref<_It2>::type> >)) constexpr mp_units::quantity<R, Rep>::quantity(FwdValue&&, R2) [with R2 = FwdValue; auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); Rep = int]’
  500 |   constexpr quantity(FwdValue&& val, R2) : quantity(::mp_units::quantity{std::forward<FwdValue>(val), R2{}})
      |             ^~~~~~~~
../../src/core/include/mp-units/framework/quantity.h:500:13: note: candidate expects 2 arguments, 1 provided
../../src/core/include/mp-units/framework/quantity.h:490:13: note: candidate 7: ‘template<class Value, class R2>  requires (Reference<R2>) && (equivalent(mp_units::quantity<R, Rep>::unit, get_unit({})) && !(RepConstructibleFrom<Rep, Value>)) constexpr mp_units::quantity<R, Rep>::quantity(Value, R2) [with R2 = Value; auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); Rep = int]’ (deleted)
  490 |   constexpr quantity(Value val, R2)
      |             ^~~~~~~~
../../src/core/include/mp-units/framework/quantity.h:490:13: note: candidate expects 2 arguments, 1 provided
../../src/core/include/mp-units/framework/quantity.h:484:13: note: candidate 8: ‘template<class R2>  requires (Reference<R2>) && equivalent(mp_units::quantity<R, Rep>::unit, get_unit({})) constexpr mp_units::quantity<R, Rep>::quantity(rep, R2) [with auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); Rep = int]’
  484 |   constexpr quantity(rep val, R2) : numerical_value_is_an_implementation_detail_(std::move(val))
      |             ^~~~~~~~
../../src/core/include/mp-units/framework/quantity.h:484:13: note: candidate expects 2 arguments, 1 provided
../../src/core/include/mp-units/framework/quantity.h:504:87: note: candidate 9: ‘constexpr mp_units::quantity<R, Rep>::quantity(rep) requires  ExplicitFromNumber<mp_units::quantity<R, Rep>::reference> [with auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); Rep = int; rep = int]’
  504 |   constexpr explicit(!mp_units::implicitly_convertible(quantity_spec, dimensionless)) quantity(rep val)
      |                                                                                       ^~~~~~~~
../../src/core/include/mp-units/framework/quantity.h:504:87: note: constraints not satisfied
../../src/core/include/mp-units/framework/quantity.h:480:3: note: candidate 10: ‘constexpr mp_units::quantity<R, Rep>::quantity() [with auto [requires mp_units::Reference<<placeholder>, >] R = mp_units::usc::degree_Fahrenheit(); Rep = int]’
  480 |   quantity() = default;
      |   ^~~~~~~~
../../src/core/include/mp-units/framework/quantity.h:480:3: note: candidate expects 0 arguments, 1 provided
../../src/core/include/mp-units/framework/quantity.h:449:7: note: candidate 11: ‘constexpr mp_units::quantity<mp_units::usc::degree_Fahrenheit(), int>::quantity(const mp_units::quantity<mp_units::usc::degree_Fahrenheit(), int>&)’
  449 | class quantity : public detail::quantity_iface {
      |       ^~~~~~~~
../../src/core/include/mp-units/framework/quantity.h:449:7: note: no known conversion for argument 1 from ‘mp_units::quantity<mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >(), int>’ to ‘const mp_units::quantity<mp_units::usc::degree_Fahrenheit(), int>&’
../../src/core/include/mp-units/framework/quantity.h:449:7: note: candidate 12: ‘constexpr mp_units::quantity<mp_units::usc::degree_Fahrenheit(), int>::quantity(mp_units::quantity<mp_units::usc::degree_Fahrenheit(), int>&&)’
../../src/core/include/mp-units/framework/quantity.h:449:7: note: no known conversion for argument 1 from ‘mp_units::quantity<mp_units::common_unit<mp_units::si::milli_<mp_units::si::kelvin>, mp_units::si::milli_<mp_units::usc::rankine> >(), int>’ to ‘mp_units::quantity<mp_units::usc::degree_Fahrenheit(), int>&&’
```
</details>

--- a/.gitignore
+++ b/.gitignore
@@ -36,6 +36,7 @@
 /.vs/
 /.vscode/
 /CMakeSettings.json
+/.qtcreator
 
 # build artifacts
 /cmake-build-*/

--- a/docs/users_guide/framework_basics/quantity_arithmetics.md
+++ b/docs/users_guide/framework_basics/quantity_arithmetics.md
@@ -341,8 +341,8 @@ if (q <= 0)
 ## Other maths
 
 This chapter scopes only on the `quantity` type's operators. However, there are many named
-math functions taking quantities as arguments. Those can be found in the _mp-units/math.h_
-header file. Among others, we can find there the following:
+math functions taking quantities or quantity points as arguments. Those can be found in the
+_mp-units/math.h_ header file. Among others, we can find there the following:
 
 - `pow()`, `sqrt()`, `cbrt()`,
 - `exp()`,
@@ -355,6 +355,10 @@ header file. Among others, we can find there the following:
 - `hypot()`,
 - `sin()`, `cos()`, `tan()`,
 - `asin()`, `acos()`, `atan()`, `atan2()`.
+
+`floor()`, `ceil()`, and `round()` take a target unit as template argument and will round
+the quantity rsp. quantity point into that unit. In the integer case `round()` rounds to
+the even value as a tie breaker.
 
 In the library, we can also find _mp-units/utility/random.h_ header file with all the
 pseudo-random number generators working on quantity types.

--- a/src/core/include/mp-units/math.h
+++ b/src/core/include/mp-units/math.h
@@ -426,6 +426,38 @@ template<Unit auto To, auto R, typename Rep>
 }
 
 /**
+ * @brief Computes the largest quantity point with integer representation and unit type To with its number not greater
+ *        than qp
+ *
+ * @param qp QuantityPoint being the base of the operation
+ * @return QuantityPoint The rounded quantity point with unit type To
+ */
+template<Unit auto To, auto R, auto PO, typename Rep>
+[[nodiscard]] constexpr QuantityPointOf<quantity_point<R, PO, Rep>::quantity_spec> auto floor(
+  const quantity_point<R, PO, Rep>& qp) noexcept
+  requires requires { qp.force_in(To); } &&
+           ((treat_as_floating_point<Rep> && (requires(Rep v) { floor(v); }
+#if MP_UNITS_HOSTED
+                                              || requires(Rep v) { std::floor(v); }
+#endif
+                                              )) ||
+            (!treat_as_floating_point<Rep> && requires { representation_values<Rep>::one(); }))
+{
+  const quantity_point res = qp.force_in(To);
+  if constexpr (treat_as_floating_point<Rep>) {
+    using ReturnType = std::remove_reference_t<decltype(res)>;
+#if MP_UNITS_HOSTED
+    using std::floor;
+#endif
+    return ReturnType{
+      quantity{static_cast<Rep>(floor(res.quantity_from_zero().numerical_value_ref_in(res.unit))), res.reference}, PO};
+  } else {
+    if (res > qp) return res - quantity{representation_values<Rep>::one(), res.reference};
+    return res;
+  }
+}
+
+/**
  * @brief Computes the smallest quantity with integer representation and unit type To with its number not less than q
  *
  * @tparam q Quantity being the base of the operation
@@ -454,6 +486,38 @@ template<Unit auto To, auto R, typename Rep>
 }
 
 /**
+ * @brief Computes the smallest quantity point with integer representation and unit type To with its number not less
+ *        than qp
+ *
+ * @param qp Quantity point being the base of the operation
+ * @return QuantityPoint The rounded quantity with unit type To
+ */
+template<Unit auto To, auto R, auto PO, typename Rep>
+[[nodiscard]] constexpr QuantityPointOf<quantity_point<R, PO, Rep>::quantity_spec> auto ceil(
+  const quantity_point<R, PO,Rep>& qp) noexcept
+  requires requires { qp.force_in(To); } &&
+           ((treat_as_floating_point<Rep> && (requires(Rep v) { ceil(v); }
+#if MP_UNITS_HOSTED
+                                              || requires(Rep v) { std::ceil(v); }
+#endif
+                                              )) ||
+            (!treat_as_floating_point<Rep> && requires { representation_values<Rep>::one(); }))
+{
+  const quantity_point res = qp.force_in(To);
+  if constexpr (treat_as_floating_point<Rep>) {
+    using ReturnType = std::remove_reference_t<decltype(res)>;
+#if MP_UNITS_HOSTED
+    using std::ceil;
+#endif
+    return ReturnType{
+      quantity{static_cast<Rep>(ceil(res.quantity_from_zero().numerical_value_ref_in(res.unit))), res.reference}, PO};
+  } else {
+    if (res < qp) return res + quantity{representation_values<Rep>::one(), res.reference};
+    return res;
+  }
+}
+
+/**
  * @brief Computes the nearest quantity with integer representation and unit type `To` to `q`
  *
  * Returns the value `res` representable in `To` unit that is the closest to `q`. If there are two
@@ -477,6 +541,37 @@ template<Unit auto To, auto R, typename Rep>
   if (diff0 == diff1) {
     // TODO How to extend this to custom representation types?
     if (static_cast<std::int64_t>(res_low.numerical_value_ref_in(To)) & 1) return res_high;
+    return res_low;
+  } else if (diff0 < diff1)
+    return res_low;
+  return res_high;
+}
+
+/**
+ * @brief Computes the nearest quantity point with integer representation and unit type `To` to `qp`
+ *
+ * Returns the value `res` representable in `To` unit that is the closest to `qp`. If there are two
+ * such values, returns the even value (that is, the value `res` such that `res % 2 == 0`).
+ *
+ * @param qp Quantity point being the base of the operation
+ * @return QuantityPoint The quantity rounded to the nearest unit `To`, rounding to even in halfway
+ *                       cases.
+ */
+template<Unit auto To, auto R, auto PO, typename Rep>
+[[nodiscard]] constexpr QuantityPointOf<quantity_point<R, PO, Rep>::quantity_spec> auto round(
+  const quantity_point<R, PO, Rep>& qp) noexcept
+  requires requires {
+    mp_units::floor<To>(qp);
+    representation_values<Rep>::one();
+  } && std::constructible_from<std::int64_t, Rep>
+{
+  const auto res_low = mp_units::floor<To>(qp);
+  const auto res_high = res_low + quantity{representation_values<Rep>::one(), res_low.reference};
+  const auto diff0 = qp - res_low;
+  const auto diff1 = res_high - qp;
+  if (diff0 == diff1) {
+    // TODO How to extend this to custom representation types?
+    if (static_cast<std::int64_t>(res_low.quantity_from_zero().numerical_value_ref_in(To)) & 1) return res_high;
     return res_low;
   } else if (diff0 < diff1)
     return res_low;

--- a/test/runtime/math_test.cpp
+++ b/test/runtime/math_test.cpp
@@ -153,50 +153,62 @@ TEST_CASE("math operations", "[math]")
     SECTION("floor 1 second with target unit second should be 1 second")
     {
       REQUIRE(floor<si::second>(1 * isq::duration[s]) == 1 * isq::duration[s]);
+      REQUIRE(floor<si::second>(point<s>(1)) == point<s>(1));
     }
     SECTION("floor 1000 milliseconds with target unit second should be 1 second")
     {
       REQUIRE(floor<si::second>(1000 * isq::duration[ms]) == 1 * isq::duration[s]);
+      REQUIRE(floor<si::second>(point<ms>(1000)) == point<s>(1));
     }
     SECTION("floor 1001 milliseconds with target unit second should be 1 second")
     {
       REQUIRE(floor<si::second>(1001 * isq::duration[ms]) == 1 * isq::duration[s]);
+      REQUIRE(floor<si::second>(point<ms>(1001)) == point<s>(1));
     }
     SECTION("floor 1999 milliseconds with target unit second should be 1 second")
     {
       REQUIRE(floor<si::second>(1999 * isq::duration[ms]) == 1 * isq::duration[s]);
+      REQUIRE(floor<si::second>(point<ms>(1999)) == point<s>(1));
     }
     SECTION("floor -1000 milliseconds with target unit second should be -1 second")
     {
       REQUIRE(floor<si::second>(-1000 * isq::duration[ms]) == -1 * isq::duration[s]);
+      REQUIRE(floor<si::second>(point<ms>(-1000)) == point<s>(-1));
     }
     SECTION("floor -999 milliseconds with target unit second should be -1 second")
     {
       REQUIRE(floor<si::second>(-999 * isq::duration[ms]) == -1 * isq::duration[s]);
+      REQUIRE(floor<si::second>(point<ms>(-999)) == point<s>(-1));
     }
     SECTION("floor 1.3 seconds with target unit second should be 1 second")
     {
       REQUIRE(floor<si::second>(1.3 * isq::duration[s]) == 1 * isq::duration[s]);
+      REQUIRE(floor<si::second>(point<s>(1.3)) == point<s>(1));
     }
     SECTION("floor -1.3 seconds with target unit second should be -2 seconds")
     {
       REQUIRE(floor<si::second>(-1.3 * isq::duration[s]) == -2 * isq::duration[s]);
+      REQUIRE(floor<si::second>(point<s>(-1.3)) == point<s>(-2));
     }
     SECTION("floor 1001. milliseconds with target unit second should be 1 second")
     {
       REQUIRE(floor<si::second>(1001. * isq::duration[ms]) == 1 * isq::duration[s]);
+      REQUIRE(floor<si::second>(point<ms>(1001.)) == point<s>(1));
     }
     SECTION("floor 1999. milliseconds with target unit second should be 1 second")
     {
       REQUIRE(floor<si::second>(1999. * isq::duration[ms]) == 1 * isq::duration[s]);
+      REQUIRE(floor<si::second>(point<ms>(1999.)) == point<s>(1));
     }
     SECTION("floor -1000. milliseconds with target unit second should be -1 second")
     {
       REQUIRE(floor<si::second>(-1000. * isq::duration[ms]) == -1 * isq::duration[s]);
+      REQUIRE(floor<si::second>(point<ms>(-1000.)) == point<s>(-1));
     }
     SECTION("floor -999. milliseconds with target unit second should be -1 second")
     {
       REQUIRE(floor<si::second>(-999. * isq::duration[ms]) == -1 * isq::duration[s]);
+      REQUIRE(floor<si::second>(point<ms>(-999.)) == point<s>(-1));
     }
 
     // TODO Add tests for `N`, `kN` and `kg * m / s2` i `kg * km / s2`
@@ -207,50 +219,62 @@ TEST_CASE("math operations", "[math]")
     SECTION("ceil 1 second with target unit second should be 1 second")
     {
       REQUIRE(ceil<si::second>(1 * isq::duration[s]) == 1 * isq::duration[s]);
+      REQUIRE(ceil<si::second>(point<s>(1)) == point<s>(1));
     }
     SECTION("ceil 1000 milliseconds with target unit second should be 1 second")
     {
       REQUIRE(ceil<si::second>(1000 * isq::duration[ms]) == 1 * isq::duration[s]);
+      REQUIRE(ceil<si::second>(point<ms>(1000)) == point<s>(1));
     }
     SECTION("ceil 1001 milliseconds with target unit second should be 2 seconds")
     {
       REQUIRE(ceil<si::second>(1001 * isq::duration[ms]) == 2 * isq::duration[s]);
+      REQUIRE(ceil<si::second>(point<ms>(1001)) == point<s>(2));
     }
     SECTION("ceil 1999 milliseconds with target unit second should be 2 seconds")
     {
       REQUIRE(ceil<si::second>(1999 * isq::duration[ms]) == 2 * isq::duration[s]);
+      REQUIRE(ceil<si::second>(point<ms>(1999)) == point<s>(2));
     }
     SECTION("ceil -1000 milliseconds with target unit second should be -1 second")
     {
       REQUIRE(ceil<si::second>(-1000 * isq::duration[ms]) == -1 * isq::duration[s]);
+      REQUIRE(ceil<si::second>(point<ms>(-1000)) == point<s>(-1));
     }
     SECTION("ceil -999 milliseconds with target unit second should be 0 seconds")
     {
       REQUIRE(ceil<si::second>(-999 * isq::duration[ms]) == 0 * isq::duration[s]);
+      REQUIRE(ceil<si::second>(point<ms>(-999)) == point<s>(0));
     }
     SECTION("ceil 1.3 seconds with target unit second should be 2 seconds")
     {
       REQUIRE(ceil<si::second>(1.3 * isq::duration[s]) == 2 * isq::duration[s]);
+      REQUIRE(ceil<si::second>(point<s>(1.3)) == point<s>(2));
     }
     SECTION("ceil -1.3 seconds with target unit second should be -1 second")
     {
       REQUIRE(ceil<si::second>(-1.3 * isq::duration[s]) == -1 * isq::duration[s]);
+      REQUIRE(ceil<si::second>(point<s>(-1.3)) == point<s>(-1));
     }
     SECTION("ceil 1001. milliseconds with target unit second should be 2 seconds")
     {
       REQUIRE(ceil<si::second>(1001. * isq::duration[ms]) == 2 * isq::duration[s]);
+      REQUIRE(ceil<si::second>(point<ms>(1001.)) == point<s>(2));
     }
     SECTION("ceil 1999. milliseconds with target unit second should be 2 seconds")
     {
       REQUIRE(ceil<si::second>(1999. * isq::duration[ms]) == 2 * isq::duration[s]);
+      REQUIRE(ceil<si::second>(point<ms>(1999.)) == point<s>(2));
     }
     SECTION("ceil -1000. milliseconds with target unit second should be -1 second")
     {
       REQUIRE(ceil<si::second>(-1000. * isq::duration[ms]) == -1 * isq::duration[s]);
+      REQUIRE(ceil<si::second>(point<ms>(-1000.)) == point<s>(-1));
     }
     SECTION("ceil -999. milliseconds with target unit second should be 0 seconds")
     {
       REQUIRE(ceil<si::second>(-999. * isq::duration[ms]) == 0 * isq::duration[s]);
+      REQUIRE(ceil<si::second>(point<ms>(-999.)) == point<s>(0));
     }
   }
 
@@ -259,86 +283,107 @@ TEST_CASE("math operations", "[math]")
     SECTION("round 1 second with target unit second should be 1 second")
     {
       REQUIRE(round<si::second>(1 * isq::duration[s]) == 1 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<s>(1)) == point<s>(1));
     }
     SECTION("round 1000 milliseconds with target unit second should be 1 second")
     {
       REQUIRE(round<si::second>(1000 * isq::duration[ms]) == 1 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(1000)) == point<s>(1));
     }
     SECTION("round 1001 milliseconds with target unit second should be 1 second")
     {
       REQUIRE(round<si::second>(1001 * isq::duration[ms]) == 1 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(1001)) == point<s>(1));
     }
     SECTION("round 1499 milliseconds with target unit second should be 1 second")
     {
       REQUIRE(round<si::second>(1499 * isq::duration[ms]) == 1 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(1499)) == point<s>(1));
     }
     SECTION("round 1500 milliseconds with target unit second should be 2 seconds")
     {
       REQUIRE(round<si::second>(1500 * isq::duration[ms]) == 2 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(1500)) == point<s>(2));
     }
     SECTION("round 1999 milliseconds with target unit second should be 2 seconds")
     {
       REQUIRE(round<si::second>(1999 * isq::duration[ms]) == 2 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(1999)) == point<s>(2));
     }
     SECTION("round -1000 milliseconds with target unit second should be -1 second")
     {
       REQUIRE(round<si::second>(-1000 * isq::duration[ms]) == -1 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(-1000)) == point<s>(-1));
     }
     SECTION("round -1001 milliseconds with target unit second should be -1 second")
     {
       REQUIRE(round<si::second>(-1001 * isq::duration[ms]) == -1 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(-1001)) == point<s>(-1));
     }
     SECTION("round -1499 milliseconds with target unit second should be -1 second")
     {
       REQUIRE(round<si::second>(-1499 * isq::duration[ms]) == -1 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(-1499)) == point<s>(-1));
     }
     SECTION("round -1500 milliseconds with target unit second should be -2 seconds")
     {
       REQUIRE(round<si::second>(-1500 * isq::duration[ms]) == -2 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(-1500)) == point<s>(-2));
     }
     SECTION("round -1999 milliseconds with target unit second should be -2 seconds")
     {
       REQUIRE(round<si::second>(-1999 * isq::duration[ms]) == -2 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(-1999)) == point<s>(-2));
     }
     SECTION("round 1000. milliseconds with target unit second should be 1 second")
     {
       REQUIRE(round<si::second>(1000. * isq::duration[ms]) == 1 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(1000.)) == point<s>(1));
     }
     SECTION("round 1001. milliseconds with target unit second should be 1 second")
     {
       REQUIRE(round<si::second>(1001. * isq::duration[ms]) == 1 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(1001.)) == point<s>(1));
     }
     SECTION("round 1499. milliseconds with target unit second should be 1 second")
     {
       REQUIRE(round<si::second>(1499. * isq::duration[ms]) == 1 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(1499.)) == point<s>(1));
     }
     SECTION("round 1500. milliseconds with target unit second should be 2 seconds")
     {
       REQUIRE(round<si::second>(1500. * isq::duration[ms]) == 2 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(1500.)) == point<s>(2));
     }
     SECTION("round 1999. milliseconds with target unit second should be 2 seconds")
     {
       REQUIRE(round<si::second>(1999. * isq::duration[ms]) == 2 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(1999.)) == point<s>(2));
     }
     SECTION("round -1000. milliseconds with target unit second should be -1 second")
     {
       REQUIRE(round<si::second>(-1000. * isq::duration[ms]) == -1 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(-1000.)) == point<s>(-1));
     }
     SECTION("round -1001. milliseconds with target unit second should be -1 second")
     {
       REQUIRE(round<si::second>(-1001. * isq::duration[ms]) == -1 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(-1001.)) == point<s>(-1));
     }
     SECTION("round -1499. milliseconds with target unit second should be -1 second")
     {
       REQUIRE(round<si::second>(-1499. * isq::duration[ms]) == -1 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(-1499.)) == point<s>(-1));
     }
     SECTION("round -1500. milliseconds with target unit second should be -2 seconds")
     {
       REQUIRE(round<si::second>(-1500. * isq::duration[ms]) == -2 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(-1500.)) == point<s>(-2));
     }
     SECTION("round -1999. milliseconds with target unit second should be -2 seconds")
     {
       REQUIRE(round<si::second>(-1999. * isq::duration[ms]) == -2 * isq::duration[s]);
+      REQUIRE(round<si::second>(point<ms>(-1999.)) == point<s>(-2));
     }
   }
 
@@ -375,6 +420,45 @@ TEST_CASE("math operations", "[math]")
     {
       REQUIRE(round<usc::degree_Fahrenheit>(delta<deg_C>(1)) ==
               delta<usc::degree_Fahrenheit>(2));  // 1 Δ°C == 1.8 Δ°F -> 2 Δ°F
+    }
+  }
+
+  SECTION("floor/ceil/round with offset units and integer representation as quantity points")
+  {
+    constexpr auto ddeg_C = si::deci<si::degree_Celsius>;  // 0.1 °C steps
+
+    SECTION("floor with a degree Fahrenheit target")
+    {
+      // -12.5 °C -> 9.5 °F => 9 °F
+      REQUIRE(floor<usc::degree_Fahrenheit>(point<ddeg_C>(-125)) == point<usc::degree_Fahrenheit>(9));
+      // -17 °C -> 1.4 °F => 1 °F
+      REQUIRE(floor<usc::degree_Fahrenheit>(point<ddeg_C>(-170)) == point<usc::degree_Fahrenheit>(1));
+      // -18 °C -> -0.4 °F => -1 °F
+      REQUIRE(floor<usc::degree_Fahrenheit>(point<ddeg_C>(-180)) == point<usc::degree_Fahrenheit>(-1));
+      // -22.5 °C -> -8.5 °F => -9 °F
+      REQUIRE(floor<usc::degree_Fahrenheit>(point<ddeg_C>(-225)) == point<usc::degree_Fahrenheit>(-9));
+    }
+    SECTION("ceil with a degree Fahrenheit target")
+    {
+      // -12.5 °C -> 9.5 °F => 10 °F
+      REQUIRE(ceil<usc::degree_Fahrenheit>(point<ddeg_C>(-125)) == point<usc::degree_Fahrenheit>(10));
+      // -17 °C -> 1.4 °F => 2 °F
+      REQUIRE(ceil<usc::degree_Fahrenheit>(point<ddeg_C>(-170)) == point<usc::degree_Fahrenheit>(2));
+      // -18 °C -> -0.4 °F => 0 °F
+      REQUIRE(ceil<usc::degree_Fahrenheit>(point<ddeg_C>(-180)) == point<usc::degree_Fahrenheit>(0));
+      // -22.5 °C -> -8.5 °F => -8 °F
+      REQUIRE(ceil<usc::degree_Fahrenheit>(point<ddeg_C>(-225)) == point<usc::degree_Fahrenheit>(-8));
+    }
+    SECTION("round with a degree Fahrenheit target (round half to even)")
+    {
+      // -12.5 °C -> 9.5 °F => 10 °F (even)
+      REQUIRE(round<usc::degree_Fahrenheit>(point<ddeg_C>(-125)) == point<usc::degree_Fahrenheit>(10));
+      // -17 °C -> 1.4 °F => 1 °F
+      REQUIRE(round<usc::degree_Fahrenheit>(point<ddeg_C>(-170)) == point<usc::degree_Fahrenheit>(1));
+      // -18 °C -> -0.4 °F => 0 °F
+      REQUIRE(round<usc::degree_Fahrenheit>(point<ddeg_C>(-180)) == point<usc::degree_Fahrenheit>(0));
+      // -22.5 °C -> -8.5 °F => -8 °F (even)
+      REQUIRE(round<usc::degree_Fahrenheit>(point<ddeg_C>(-225)) == point<usc::degree_Fahrenheit>(-8));
     }
   }
 

